src/Host/Broker/Impl/Pipes/MessagePipe.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.Extensions.Logging;
+using Microsoft.R.Host.Client;
+
+namespace Microsoft.R.Host.Broker.Pipes {
+    public class MessagePipe {
+        private readonly ILogger _logger;
+        private readonly BufferBlock<byte[]> _hostMessages = new BufferBlock<byte[]>();
+        private readonly BufferBlock<byte[]> _clientMessages = new BufferBlock<byte[]>();
+
+        private ConcurrentDictionary<ulong, byte[]> _sentPendingRequests = new ConcurrentDictionary<ulong, byte[]>();
+        private Queue<byte[]> _unsentPendingRequests = new Queue<byte[]>();
+
+        private byte[] _handshake;
+        private IMessagePipeEnd _hostEnd, _clientEnd;
+
+        private abstract class PipeEnd : IMessagePipeEnd {
```

Comment:
... on the other hand, if pipe implements IMessagePipeEnd directly, access to it may be gained directly, and multiple hosts might be attached to the same pipe by mistake. Right now they all have to go through ConnectHost, which enforces a single end (obviously someone can call the method once and then share the returned value, but I think it's less likely to happen accidentally).

---
In reply to: [74839548](https://github.com/Microsoft/RTVS/pull/2173#discussion_r74839548) [](ancestors = 74839548,74839003,74838114,74837833,74833662,74833429)